### PR TITLE
Add ability to show/hide menu items

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#881

I tested this by running the example program on Windows 10 64-bit, OS X, and Ubuntu 14.04.5 64-bit.

### CODE REVIEW

File: systray/systray/ReadMe.txt [hunk 1]
```diff
@@ -1,48 +0,0 @@
-========================================================================
```
Comment (oxtoacart) [inline systray/systray/ReadMe.txt]: This and a lot of the below files are code that shouldn't have been checked in.

File: systray.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ var (
 	menuItems     = make(map[int32]*MenuItem)
 	menuItemsLock sync.RWMutex
 
-	currentID int32
+	currentID = int32(-1)
```
Comment (oxtoacart) [inline systray.go]: Starting at -1 makes the id and the position of the menu item line up, which is needed on Windows.

File: systray/systray/systray.cpp [hunk 3]
```diff
@@ -41,22 +41,11 @@ void ShowMenu(HWND hWnd) {
...
-	return menuItemInfo.dwItemData;
-}
-
 LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam) {
 	switch (message) {
-		case WM_MENUCOMMAND:
+		case WM_COMMAND:
 			{
-				int menuId = GetMenuItemId(wParam);
+				int menuId = LOWORD(wParam);
```
Comment (oxtoacart) [inline systray/systray/systray.cpp]: We now manage items by ID rather than position.

File: systray/systray/systray.cpp [hunk 4]
```diff
@@ -204,23 +192,16 @@ void add_or_update_menu_item(int menuId, wchar_t* title, wchar_t* tooltip, short
...
-	if (i == itemCount) {
-		InsertMenuItem(hTrayMenu, -1, TRUE, &menuItemInfo);
+	BOOL setOK = SetMenuItemInfo(hTrayMenu, menuId, FALSE, &menuItemInfo);
+	if (!setOK) {
+		InsertMenuItem(hTrayMenu, menuId, TRUE, &menuItemInfo);
 	}
 }
 
+void hide_menu_item(int menuId) {
+	DeleteMenu(hTrayMenu, menuId, MF_BYCOMMAND);
```
Comment (oxtoacart) [inline systray/systray/systray.cpp]: To hide on Windows, we delete.

File: systray_linux.c [hunk 5]
```diff
@@ -110,6 +110,32 @@ gboolean do_add_or_update_menu_item(gpointer data) {
...
+gboolean do_hide_menu_item(gpointer data) {
+	MenuItemInfo *mii = (MenuItemInfo*)data;
+	GList* it;
+	for(it = global_menu_items; it != NULL; it = it->next) {
+		MenuItemNode* item = (MenuItemNode*)(it->data);
+		if(item->menu_id == mii->menu_id){
+			gtk_widget_hide(GTK_WIDGET(item->menu_item));
+			return;
+		}
+	}
```
Comment (joesis) [inline systray_linux.c]: Should always return FALSE. The implicit return value is undefined.

File: systray_linux.c [hunk 6]
```diff
@@ -110,6 +110,32 @@ gboolean do_add_or_update_menu_item(gpointer data) {
...
+gboolean do_show_menu_item(gpointer data) {
+	MenuItemInfo *mii = (MenuItemInfo*)data;
+	GList* it;
+	for(it = global_menu_items; it != NULL; it = it->next) {
+		MenuItemNode* item = (MenuItemNode*)(it->data);
+		if(item->menu_id == mii->menu_id){
+			gtk_widget_show(GTK_WIDGET(item->menu_item));
+			return;
+		}
+	}
```
Comment (joesis) [inline systray_linux.c]: Should always return FALSE. The implicit return value is undefined.

File: systray_darwin.m [hunk 7]
```diff
@@ -108,6 +108,28 @@ - (void) add_or_update_menu_item:(MenuItem*) item
...
+
+- (void) show_menu_item:(NSNumber*) menuId
+{
+  NSMenuItem* menuItem;
+  int existedMenuIndex = [menu indexOfItemWithRepresentedObject: menuId];
+  if (existedMenuIndex == -1) {
+    return;
+  }
+  menuItem = [menu itemAtIndex: existedMenuIndex];
+  [menuItem setHidden:NO];
```
Comment (joesis) [inline systray_darwin.m]: There's no difference in practice, but maybe use `FALSE` for symmetry?

File: systray/systray/systray.cpp (hunk 3 above)
Comment (joesis) [reply systray/systray/systray.cpp]: I'm curious why I didn't use `wID` field in the first place? There was probably some reason, but the change looks good!

File: systray/systray/systray.cpp [hunk 8]
```diff
@@ -204,23 +192,16 @@ void add_or_update_menu_item(int menuId, wchar_t* title, wchar_t* tooltip, short
...
-		if (menuId == id) {
-			SetMenuItemInfo(hTrayMenu, i, TRUE, &menuItemInfo);
-			break;
-		}
-	}
-	if (i == itemCount) {
-		InsertMenuItem(hTrayMenu, -1, TRUE, &menuItemInfo);
+	BOOL setOK = SetMenuItemInfo(hTrayMenu, menuId, FALSE, &menuItemInfo);
+	if (!setOK) {
+		InsertMenuItem(hTrayMenu, menuId, TRUE, &menuItemInfo);
```
Comment (joesis) [inline systray/systray/systray.cpp]: The 3rd param should be `FALSE` I guess?

**Comment by joesis** [comment]
The DLLs didn't export `hide_menu_item` function. Running the example causes "Failed to find hide_menu_item procedure" panic too. I guess they weren't compiled from the latest code?

File: systray/systray/systray.cpp (hunk 8 above)
Comment (oxtoacart) [reply systray/systray/systray.cpp]: Actually no, we're using the menuId as the position here to make sure that when we show the menu item after hiding it, it goes back in the right position.

File: systray_darwin.m (hunk 7 above)
Comment (oxtoacart) [reply systray_darwin.m]: Fixed.

File: systray_linux.c (hunk 6 above)
Comment (oxtoacart) [reply systray_linux.c]: Fixed.

File: systray_linux.c (hunk 5 above)
Comment (oxtoacart) [reply systray_linux.c]: Fixed.

**Comment by oxtoacart** [comment]
@joesis Sorry about that, I fixed the DLLs. I was having some git issues last night, so they go overwritten. Example should work now.

File: systray/systray/systray.cpp (hunk 8 above)
Comment (joesis) [reply systray/systray/systray.cpp]: Gotcha

Comment (joesis) [reply systray/systray/systray.cpp]: Maybe worth a comment?

Comment (oxtoacart) [reply systray/systray/systray.cpp]: Good idea, done.

**Comment by joesis** [comment]
Cool @oxtoacart . LGTM!

**Comment by oxtoacart** [comment]
@joesis Thanks for the review!

**Pull request merged**